EQL: Use In from QL

rw-access opened the pull request:
Resolves #52939 

Using `In` directly now, instead of creating an OR of Equals.

elasticmachine commented:
Pinging @elastic/es-search (:Search/EQL)

rw-access commented:
Ah. Yeah I can improve the tests. There were more tests for EQL than just the test_queries.toml that tested parsing, optimization, etc.  For example, https://github.com/endgameinc/eql/blob/master/tests/test_optimizations.py#L11-L109

I don't think we have to pull each one of those over, but I can add tests for these as well. At what level do  you think it makes the most sense to test? Parser? Optimizer? Or adding examples to `test_queries.toml` where we test end-to-end?

costin commented:
> Parser? Optimizer? Or adding examples to test_queries.toml where we test end-to-end?

Since this is mainly for execution, I wouldn't worry to much about the `Optimizer` and add a few that check the generated query and few end-to-end.
Those would automatically pick up on the parser, analyzer and optimizer.
In general, we strive for unit tests to check as much of the desired behavior in a fast fashion and then check the mix in the end-to-end which, while comprehensive, it's slower and harder to debug in case of failures.

The unit tests are typically written after the feature is completed to make sure it does what we want while the integration tests after the former passes to make sure our assumptions across the whole chain are correct.

rw-access commented:
I added a few more unit tests.
* `In` requires at leasts one value, otherwise it's invalid syntax. Added a test for that behavior
* Added a unit test with numeric descending values
* Added a unit test for duplicate values
* Updated tests to more explicitly verify the AST
* Added a test for `In` with mixed fields

costin commented:
The bwc indicates that the master has changed.

@elasticmachine update branch

costin commented:
@elasticmachine update branch

costin commented on x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/planner/QueryFolderOkTests.java:
```diff
@@ -45,18 +45,13 @@
-                            "\"term\":{\"opcode\":{\"value\":1",
-                            "\"term\":{\"opcode\":{\"value\":2",
-                            "\"term\":{\"opcode\":{\"value\":3",
+                            "{\"terms\":{\"opcode\":[0,1,2,3],\"boost\":1.0}}],\"boost\":1.0}}",
```
I would just check the terms and opcode not the boost - that's a default we don't influence.

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/planner/ExpressionTranslators.java:
```diff
@@ -41,28 +42,36 @@
             new BinaryLogic(),
             new Nots(),
             new Likes(),
+            new InComparisons(),
```
What's the behavior of wildcard string in `In` - are they supported or not? Should we check their existence and if we discover some, throw an error? Or better yet convert them into `Equals` and `And`.
I can see how somebody doing an equals against two patterns might be tempted to use an `In` with both of them instead.

rw-access replied:
great question.
`In` is always exact match (+/- case-sensitivity for ES).
we use the wildcard function when matching multiple patterns: `wildcard(command_line, "*localgroup*", "*group*", "*user*")`.

there has been a little confusion by some users that have used `in` with wildcards, so we can change that behavior if that reduces friction. I've avoided that so far, because it would quietly change the behavior of existing syntax. But eventually it may make sense to switch.

costin commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/planner/ExpressionTranslators.java:
```diff
@@ -314,6 +323,36 @@ public static Query doTranslate(Range r, TranslatorHandler handler) {
         }
     }
 
+    public static class InComparisons extends ExpressionTranslator<In> {
```
If you're comfortable, please update the rule in SQL as well similar to `Ranges` or `BinaryLogic` - essentially extend this one from QL and add the if/else block with `onAggs`.

rw-access commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryTranslator.java:
```diff
@@ -461,23 +456,7 @@ protected QueryTranslation asQuery(In in, boolean onAggs, TranslatorHandler hand
-                    DataType dt = list.get(0).dataType();
-                    Set<Object> set = new LinkedHashSet<>(CollectionUtils.mapSize(list.size()));
-                    for (Expression e : list) {
-                        set.add(SqlDataTypeConverter.convert(e.fold(), dt));
```
One thing to note @costin is that `SqlDataTypeConverter` isn't available in QL,  so this is now just `valueOf`. Looks like SQL tests are still passing though
https://github.com/elastic/elasticsearch/blob/d6757d01e376a233c45537814452b43d7ff84b94/x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/planner/ExpressionTranslators.java#L345

costin replied:
This can addressed by adding a `convertTo()` method which by default uses `DataTypeConverter` inside QL and `SqlDataTypeConverter` inside SQL.
That said, if the tests are passing across the board I would leave this as is to avoid doing any implicit conversion.
Potentially add a comment mentioning that before there was a conversion happening to help track this down in case a weird bug occurs in the future.

costin replied:
Looks like the failing tests are from the SQL tests so the solution would be to keep the extend `converTo` in the SQL `ExpressionTranslator` and use `SqlDataTypeConverter` there as before.

rw-access replied:
Looking some more, I see this method that could also work. Could just make it `public` instead of `protected` and use this:
https://github.com/elastic/elasticsearch/blob/7425042f86f66df8c207c5e96f9b9848bda2b4c3/x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/predicate/operator/comparison/In.java#L106-L112

And since In is subclassed in SQL, this has an override for SqlDataTypeConverter
https://github.com/elastic/elasticsearch/blob/7425042f86f66df8c207c5e96f9b9848bda2b4c3/x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/predicate/operator/comparison/In.java#L106-L112

costin replied:
I think you're suggesting to use the `foldAndConvertList` from `In` outside of it.
We're trying to prevent that since when we expanded the types, it was discovered that there were quite a number of implicit conversions which are simply bad practices since they appear "magic" and allow lenient behavior.
We actually had a `Foldables` class that did that but it wasn't used anywhere outside `In` hence why it got moved there.
If we are to make it public and end up using it, we're back where we started.

This method might fix the current sets of tests however I expect more `ExpressionTranslation` rule in the future to require the `valueOf` method and thus delegation to the appropriate converter for their respective projects.

rw-access commented on x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/planner/QueryFolderFailTests.java:
```diff
@@ -17,9 +18,10 @@ public void testPropertyEquationFilterUnsupported() {
 
     public void testPropertyEquationInClauseFilterUnsupported() {
-        QlIllegalArgumentException e = expectThrows(QlIllegalArgumentException.class,
+        VerificationException e = expectThrows(VerificationException.class,
```
I noticed that this exception changed. Should this also be the case for `testPropertyEquationFilterUnsupported`?
https://github.com/elastic/elasticsearch/pull/53244/files#diff-3132b6847876733b6fd2161331f9d044L13-R17

costin replied:
Not sure I follow - the other test doesn't have a `in` query so I don't think it should.

rw-access commented on x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/planner/ExpressionTranslators.java:
```diff
@@ -314,6 +323,36 @@ public static Query doTranslate(Range r, TranslatorHandler handler) {
+                Set<Object> set = new LinkedHashSet<>(CollectionUtils.mapSize(list.size()));
+
+                for (Expression e : list) {
+                    set.add(valueOf(e));
```
Behavior slightly changed. See comment in SQL
https://github.com/elastic/elasticsearch/pull/53244/files#r394547453

matriv commented:
@elastic/es-ql
